src/cluster_linearize.h

```diff
@@ -723,19 +729,25 @@ class SearchCandidateFinder
          *
          * - inc: the "inc" value for the new work item (must be topological).
          * - und: the "und" value for the new work item ((inc | und) must be topological).
+         * - pot: a subset of the "pot" value for the new work item (but a superset of inc).
+         *        It does not need to be the full pot value; missing pot transactions will be added
+         *        to it by add_fn.
+         * - grow_inc: whether to attempt moving transactions from und to inc, if it can be proven
```

Comment:
I think this has helped a lot, "clusterlin: avoid recomputing potential set on every split (optimization)" now makes sense to me. Lots of interlocking parts, so having small bite-sizes helps.

Can you motivate why `reconsider_pot` is set as true on inclusion of `anc` and not on exclude? 

When `anc` is included, it's included in both `inc` and `pot`, then:

`if (reconsider_pot) consider_inc = pot.transactions - inc.transactions;`

so no part of `anc` will be in `consider_inc`.

Reply:
@sipa thanks this is really helpful, particularly this part:

> 4. anc(t) will always include at least one UL transaction (*)
> (*) The jump ahead is exactly what guarantees that anc(t) always includes a UL transaction. If anc(t) was entirely within UH, it would be topologically valid, and thus the parent work item's jump ahead step would have moved it to I already.

However I'm still having trouble with this part:

> Compare this to what happens in an exclusion branch ...
> 4. desc(t) may include a UH transaction, and because of that it is possible that more transactions can be moved from UL to UH.
> 5. Because of the above, it is possible that there are new topologically-valid subsets of I+UH now that can be moved to I. This can only be due to UH entries added in (4).

It seems true that only the entries added in (4) can cause jump-aheads. However, I'm not convinced that existing `pot` transactions cannot be jump-aheadable. For instance, it's possible that we added a new tx that is the parent of a tx that was already in `pot`.

I added an `assert(anc.Overlaps(elem.und - elem.pot.transactions))` in `split_fn`, right before we call `add_fn` for the inclusion branch in order to check that "jump ahead guarantees that anc(t) always includes a UL transaction". The assertion hit, and essentially what I saw was:

1. Going into a split, we had a really high pot feerate, which included tx `C`.
2. For the exclusion branch, desc(t) included lots of high feerate transactions which lowered the pot feerate. `C` is still in pot. At the start of this `add_fn`, we add tx `P` to pot, since its feerate is now high enough.
3. `P` is part of `consider_inc`, and is added to `inc` in the jump-ahead.
4. `C`, which a child of `P`, would be good because anc(C) is a topological subset of `pot`, but is skipped because `reconsider_pot=false`
5. Later, we split on `C`. `C` is part of UH, not UL, which hits the assert. anc(C) = C which is in UH, but it wasn't added to `inc` during a jump-ahead. This is because `C` was added to UH when its ancestor `P` was considered UL, and `reconsider_pot=false` when `P` was added to UH.

Does this make sense? Or maybe my `assert` isn't the right interpretation of this? I don't think this changes what `FindCandidateSet` outputs, but it means I still don't quite understand the " avoid jump ahead in unmodified pot sets" commit :(

(fwiw the fuzz I crashed on is `echo "AQBJCAgAIysjIyMiIyMjIyMjIyMjIyM6IyMIACMrIyMjIiMjIyMjIyMjIyMjIyMjCBlE+xEvRG9v
AggAbwAR" | base 64 -d`)

Reply:
How significant do we expect the `reconsider_pot` optimization to be? Locally my `LinearizePerIter*TxWorstCase` bench doesn't seem to do better after c68d26a0e97.

Reply:
Here's what I get on 255571339e7:
```
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|          724,714.00 |            1,379.85 |    2.4% |      0.01 | `Linearize32TxWorstCase15000Iters`
|          245,712.75 |            4,069.79 |    0.3% |      0.01 | `Linearize32TxWorstCase5000Iters`
|          688,589.50 |            1,452.24 |    5.2% |      0.01 | :wavy_dash: `Linearize48TxWorstCase15000Iters` (Unstable with ~1.2 iters. Increase `minEpochIterations` to e.g. 12)
|          233,424.25 |            4,284.05 |    4.8% |      0.01 | `Linearize48TxWorstCase5000Iters`
|          662,959.00 |            1,508.39 |    1.7% |      0.01 | `Linearize64TxWorstCase15000Iters`
|          242,351.75 |            4,126.23 |    1.9% |      0.01 | `Linearize64TxWorstCase5000Iters`
|        1,051,925.00 |              950.64 |    4.3% |      0.01 | `Linearize75TxWorstCase15000Iters`
|          344,038.00 |            2,906.66 |    1.8% |      0.01 | `Linearize75TxWorstCase5000Iters`
|        1,023,533.00 |              977.01 |    2.1% |      0.01 | `Linearize99TxWorstCase15000Iters`
|          356,557.67 |            2,804.60 |    1.3% |      0.01 | `Linearize99TxWorstCase5000Iters`
```
on d04997871f5:
```
|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|          705,210.00 |            1,418.02 |    4.4% |      0.01 | `Linearize32TxWorstCase15000Iters`
|          231,874.00 |            4,312.69 |    0.2% |      0.01 | `Linearize32TxWorstCase5000Iters`
|          664,206.00 |            1,505.56 |    4.7% |      0.01 | `Linearize48TxWorstCase15000Iters`
|          222,808.25 |            4,488.16 |    1.9% |      0.01 | `Linearize48TxWorstCase5000Iters`
|          627,877.00 |            1,592.67 |    0.2% |      0.01 | `Linearize64TxWorstCase15000Iters`
|          223,255.50 |            4,479.17 |    2.6% |      0.01 | `Linearize64TxWorstCase5000Iters`
|          993,733.00 |            1,006.31 |    3.8% |      0.01 | `Linearize75TxWorstCase15000Iters`
|          327,429.67 |            3,054.09 |    2.3% |      0.01 | `Linearize75TxWorstCase5000Iters`
|          957,057.00 |            1,044.87 |    3.9% |      0.01 | `Linearize99TxWorstCase15000Iters`
|          339,669.33 |            2,944.04 |    1.8% |      0.01 | `Linearize99TxWorstCase5000Iters`
```

Looks somewhat better but not much / close to the err%:

| Benchmark                            | Table1 ns/op | Table2 ns/op | Percentage Speedup |
|--------------------------------------|--------------|--------------|--------------------|
| `Linearize32TxWorstCase15000Iters`   | 724,714.00   | 705,210.00   | 2.70%              |
| `Linearize32TxWorstCase5000Iters`    | 245,712.75   | 231,874.00   | 5.64%              |
| `Linearize48TxWorstCase15000Iters`   | 688,589.50   | 664,206.00   | 3.54%              |
| `Linearize48TxWorstCase5000Iters`    | 233,424.25   | 222,808.25   | 4.55%              |
| `Linearize64TxWorstCase15000Iters`   | 662,959.00   | 627,877.00   | 5.30%              |
| `Linearize64TxWorstCase5000Iters`    | 242,351.75   | 223,255.50   | 7.88%              |
| `Linearize75TxWorstCase15000Iters`   | 1,051,925.00 | 993,733.00   | 5.54%              |
| `Linearize75TxWorstCase5000Iters`    | 344,038.00   | 327,429.67   | 4.84%              |
| `Linearize99TxWorstCase15000Iters`   | 1,023,533.00 | 957,057.00   | 6.49%              |
| `Linearize99TxWorstCase5000Iters`    | 356,557.67   | 339,669.33   | 4.74%              |